File: typescript/src/p_0876_middle_node.ts
/**
 * Definition for singly-linked list.
 */

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}


function middleNode(head: ListNode | null): ListNode | null {
    if (!head) return null;

    let slow: ListNode | null = head;
    let fast: ListNode | null = head;

    while (fast && fast.next) {
        slow = slow!.next;
        fast = fast.next.next;
    }

    return slow;
};

function tests() {
    let head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    let expected = 3;
    let result = middleNode(head);
    console.assert(result?.val === expected, `Test case 1 failed: middleNode(${head}) = ${result?.val}, expected ${expected}`);
    console.log(`Test case 1 passed: middleNode(${head}) = ${result?.val}`);

    head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, new ListNode(6))))));
    expected = 4;
    result = middleNode(head);
    console.assert(result?.val === expected, `Test case 2 failed: middleNode(${head}) = ${result?.val}, expected ${expected}`);
    console.log(`Test case 2 passed: middleNode(${head}) = ${result?.val}`);

    head = new ListNode(1);
    expected = 1;
    result = middleNode(head);
    console.assert(result?.val === expected, `Test case 3 failed: middleNode(${head}) = ${result?.val}, expected ${expected}`);
    console.log(`Test case 3 passed: middleNode(${head}) = ${result?.val}`);
}

tests();

File: python/src/p_0876_middle_node.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

if __name__ == '__main__':
    s = Solution()

    # Example 1
    n = [1, 2, 3, 4, 5]
    head = ListNode(n[0], ListNode(n[1], ListNode(n[2], ListNode(n[3], ListNode(n[4])))))
    print(f'list={n}')
    mid = s.middleNode(head)
    print(f'middle node={mid.val}')
    assert mid.val == 3
    print('-' * 10)

    # Example 2
    n = [1, 2, 3, 4, 5, 6]
    head = ListNode(n[0], ListNode(n[1], ListNode(n[2], ListNode(n[3], ListNode(n[4], ListNode(n[5]))))))
    print(f'list={n}')
    mid = s.middleNode(head)
    print(f'middle node={mid.val}')
    assert mid.val == 4
    print('-' * 10)

    # Example 3
    n = [1]
    head = ListNode(n[0])
    print(f'list={n}')
    mid = s.middleNode(head)
    print(f'middle node={mid.val}')
    assert mid.val == 1
    print('-' * 10)

    # Example 4
    n = [1, 2]
    head = ListNode(n[0], ListNode(n[1]))
    print(f'list={n}')
    mid = s.middleNode(head)
    print(f'middle node={mid.val}')
    assert mid.val == 2
    print('-' * 10)
